x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/WaitForIndexingCompleteStep.java

```diff
@@ -0,0 +1,93 @@
+package org.elasticsearch.xpack.core.indexlifecycle;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.index.Index;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.elasticsearch.xpack.core.indexlifecycle.UnfollowAction.CCR_METADATA_KEY;
+
+final class WaitForIndexingCompleteStep extends ClusterStateWaitStep {
+    private static final Logger logger = LogManager.getLogger(WaitForIndexingCompleteStep.class);
+
+    static final String NAME = "wait-for-indexing-complete";
+
+    WaitForIndexingCompleteStep(StepKey key, StepKey nextStepKey) {
+        super(key, nextStepKey);
+    }
+
+    @Override
+    public Result isConditionMet(Index index, ClusterState clusterState) {
+        IndexMetaData followerIndex = clusterState.metaData().index(index);
+        if (followerIndex == null) {
+            // Index must have been since deleted, ignore it
+            logger.debug("[{}] lifecycle action for index [{}] executed but index no longer exists", getKey().getAction(), index.getName());
+            return new Result(false, null);
+        }
+        Map<String, String> customIndexMetadata = followerIndex.getCustomData(CCR_METADATA_KEY);
+        if (customIndexMetadata == null) {
+            return new Result(true, null);
+        }
+
+        boolean indexingComplete = LifecycleSettings.LIFECYCLE_INDEXING_COMPLETE_SETTING.get(followerIndex.getSettings());
+        if (indexingComplete) {
+            return new Result(true, null);
+        } else {
+            return new Result(false, new Info(indexingComplete));
+        }
+    }
+
+    static final class Info implements ToXContentObject {
+
+        static final ParseField MESSAGE_FIELD = new ParseField("message");
+        static final ParseField INDEXING_COMPLETE = new ParseField(LifecycleSettings.LIFECYCLE_INDEXING_COMPLETE);
+
+        private final String message;
+        private final boolean indexingComplete;
+
+        Info(boolean indexingComplete) {
+            this.indexingComplete = indexingComplete;
+            this.message = "waiting for the [" + LifecycleSettings.LIFECYCLE_INDEXING_COMPLETE +
+                "] setting to be set to true on the leader index, it is currently [" + indexingComplete + "]";
+        }
+
+        String getMessage() {
+            return message;
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
+            builder.field(MESSAGE_FIELD.getPreferredName(), message);
+            builder.field(INDEXING_COMPLETE.getPreferredName(), indexingComplete);
```

Comment:
We could probably just hardcode `false` here? The class could be called `IndexingNotCompleteInfo` then too, since it's never used other than when the indexing is not complete

Reply:
Fair, I went back and forth a little on whether to hardcode false or pass in the actual value. I'll switch it.